### ISSUE TITLE
[RFC]: extend MoyaBot to automatically follow up with issues & PRs where we're waiting for author response

### ISSUE THREAD

**Issue by AndrewSB** [Moya/Moya#1071]
I recently saw @KrauseFx's talk on scaling open source communities (great talk, I'd recommend it to everyone who's doing anything at all in open source). https://news.realm.io/news/tryswift-felix-krause-scaling-open-source-communities-github-management

One of my big take aways was fastlane's fastlane-bot. Which would automatically follow up with issues and close them if there was no response.
Right now, a couple of contributors have been manually doing that, but I think it would be awesome if we could add some of fastlane-bot's functionality to our bot so we don't have to politely close stale issues

Any thoughts from other contributors? Does anyone want to take a shot at implementing that for MoyaBot?

**Comment by BasThomas** [comment]
I really like this - though looking at fastlane-bot's [repository](https://github.com/fastlane/issue-bot), I have no idea how it would work implementing it. Does anybody have an idea, or should we go ahead and file an issue there?

**Comment by sunshinejr** [comment]
We can wait for [Peril](https://github.com/danger/peril) from @orta. It looks really promising. It is not fully stable yet, but it can already post on issues AFAIK.

**Comment by AndrewSB** [comment]
Peril does look really nice, let's wait for that to be usable

**Comment by pedrovereza** [comment]
Just came across this: https://github.com/probot/stale 🤔

**Comment by sunshinejr** [comment]
This one is really cool! I think we can try to do it. What do you think, @Moya/contributors?

**Comment by BasThomas** [comment]
Sure, this would be really nice! Re: MoyaBot: I think you should ping @ashfurrow to ask if that's possible. :)

**Comment by orta** [comment]
We just merged it in on CocoaPods - https://github.com/CocoaPods/CocoaPods/pull/6745 - I can't really compete with GitHub on resources for this kind of project, so I wouldn't wait on peril

**Comment by sunshinejr** [comment]
Ye, and this one is looking cool. The question is if we want/can use MoyaBot for that. Also, we would need someone to do a PR with config. The values we will probably discuss in a PR (if they are off).

**Comment by AndrewSB** [comment]
😆
I'm happy this is happening!

**Comment by sunshinejr** [comment]
I was talking with Ash and we've figured that MoyaBot would need a write access so it would be better to use different bot or someone else's access token. Thus, it might be better just to install this stale bot and go from that. I'm gonna do a PR with some default values for config so we can discuss it there 👍

**Comment by BasThomas** [comment]
Awesome! 👊

**Comment by sunshinejr** [comment]
I've opened a PR#1101 with the config, please let me know what do you guys think!

What we need to get this bot running:
- [ ] #1101 merged.
- [ ] Re-visit issues that might have need a label to not flag it.
- [ ] Install the bot & iterate with the feedback.

Is there something I didn't think about? 🤔

**Comment by AndrewSB** [comment]
That sounds good to me. Thanks for figuring this out!!

On to-do 2, let's wait to add labels until we've iterated on the bot, just
in case we end up not liking it

**Comment by sunshinejr** [comment]
I've revisited some issues anyways - I've flagged most of documentation ones to leave it as it is important that we have these open and add it in the future. If you guys want to make last review, I'm gonna try to launch the bot tomorrow morning or so.

**Comment by pedrovereza** [comment]
@sunshinejr thanks for taking the time! ❤️

**Comment by sunshinejr** [comment]
Bot added - let see how it goes :D

**Comment by sunshinejr** [comment]
Wanted to make a quick summary after a month, but seems like our bot is stuck? Maybe wrong configuration? Anyone sees something strange [here](https://github.com/Moya/Moya/blob/master/.github/stale.yml)? Because in issue like #1056, label was added 28 days ago, which means 14 days ago it should be closed, but it is still open 🤔

**Comment by SD10** [comment]
@sunshinejr The configuration looks correct to me. Do you think its a permission issue considering it can comment on issues but not close them?

**Comment by sunshinejr** [comment]
👍 nice catch, @SD10. Can you take care of that issue and see if this PR helped?

**Comment by SD10** [comment]
@sunshinejr Thanks for checking the permissions. This change is already merged in. It may take some time to determine if this changes anything. There's little and no official documentation on _when_ probot will scan the issues. Some info [here](https://github.com/probot/stale/pull/41) if anyone is interested.

I'll continue watching this until it's resolved.

**Comment by sunshinejr** [comment]
Great, thank you Steven! 🍻

**Comment by SD10** [comment]
I think the problem is with our label having spaces. 
The method used to grab stale issues at probot/stale doesn't wrap the label in quotes when performing the search query to find closable issues. I created a [PR](https://github.com/probot/stale/pull/44) at stale to fix this (if it is the problem 😅).

**Comment by pedrovereza** [comment]
@SD10 That's exactly what I was testing https://github.com/pedrovereza/stale-bot-test/issues/2 😄

**Comment by SD10** [comment]
@pedrovereza You read my mind! I was going to ask you to please test this there 👍

**Comment by SD10** [comment]
@pedrovereza What do you think about changing our label to one without spaces? The support at stale looks limited. I don't think they'll be merging that PR anytime soon.

**Comment by pedrovereza** [comment]
@SD10 I'm ok waiting a bit longer for them to take a look at the PR. But if we decide to change the label, we need to revisit [all the ones currently marked by the bot](https://github.com/Moya/Moya/issues?q=is%3Aissue+is%3Aopen+label%3A%22no+recent+activity%22).

@sunshinejr @AndrewSB @BasThomas thoughts? 😄

**Comment by sunshinejr** [comment]
This should be as easy as changing the name of one label & updating the StaleBot config, correct? If so, and if its gonna fix the bot, I am up for the change 👍 I also don't mind waiting few days before we make a change, but seeing from the Issues/PRs it might be quite some time before they merge it.

**Comment by sunshinejr** [comment]
This should be as easy as changing the name of one label & updating the StaleBot config, correct? If so, and if its gonna fix the bot, I am up for the change 👍 I also don't mind waiting few days before we make a change, but seeing from the Issues/PRs it might be quite some time before they merge it.

**Comment by SD10** [comment]
@sunshinejr We would probably have to resolve all the current issues manually. 

Updating a label is an action that will update the last modified date of an issue. Issues that are currently marked as `no recent activity` won't get marked with the new label until 45 days after they were marked with the first label 😐

It would work immediately for all issues not already marked.

**Comment by sunshinejr** [comment]
But if we rename the label to something without spaces, then the update date shouldn't change (since it is the same tag object but a different name) and if we change name of the label in the config, the bot should catch that one as well, right? Or am I missing something?

**Comment by BasThomas** [comment]
The API has a "[last updated at](https://developer.github.com/v3/issues/#list-issues-for-a-repository)". This *might* (?) be updated when a label changes? I'm not sure.

**Comment by pedrovereza** [comment]
We can try changing the label name and bot config, worst case we'll have to close them manually (which was the original plan anyway) 😄

**Comment by sunshinejr** [comment]
Well, we can (probably) make sure it will work! 😄 

I've just tested it on my [test repo](https://github.com/sunshinejr/IssuesTest). My steps to reproduce:
1. Create a new label with space, `"new label"`
2. Create an issue with this label.
3. Check `updated_at` field (since from [here](https://github.com/probot/stale/blob/master/lib/stale.js#L41) I understand that it queries based on `updated_at` field). Date: `2017-07-10T15:05:36Z`
4. Rename the label to a label without space, `"new_label"`.
5. Check `updated_at` field again. Date: `2017-07-10T15:05:36Z`, which is the same as before.

_(Note: I was using this [request](https://api.github.com/repos/sunshinejr/issuesTest/issues) to test it.)_

So if the data is correct and I didn't miss anything, whenever we change the label, the `update_at` field of the issue is not updated. Therefore, we can assume we won't need to take care of all the issues above. 

That was fun! 😄

**Comment by SD10** [comment]
I actually think it does the query using the `updated` search qualifier:
i.e) `is:issue is:open updated:>2017-07-10`

I think that PR should've taken effect immediately, so I'm thinking there's a stateful issue.
Regardless, I think that was the right step so let's see how the next marked issue is handled.

**Comment by SD10** [comment]
It looks like probot has stopped slacking off 😅. 
I don't know if we want to keep this issue open to discuss the merit of this automation tool or close it now that the bot has completed a full cycle.

**Comment by pedrovereza** [comment]
I'm up for closing this issue now that the bot is finally working :tada:

We originally wanted @MoyaBot to do the job, but I don't think the effort to deploy our own bot is worth it.

@MoyaBot: nothing personal ❤️

**Comment by sunshinejr** [comment]
The issue was created to integrate the bot, and because it is both integrated and working now, there is a reason to close it 😄 We can create another issue for things related to this bot if something pops up.  

Thanks everyone for feedback & help! 🎉

**Comment by BasThomas** [comment]
I saw that over at Fastlane, there's actually another step locking an issue some time after an issue has been closed. Would it be possible to add this as well? And then maybe add another comment before locking that asks to create a new issue?

**Comment by sunshinejr** [comment]
Is that also using probot/stale, @BasThomas? I think we had this conversation here at #1101 and it wasn't possible, but using another plugin could work (though I don't know if we can do it without deploying it).
